src/main/java/com/gooddata/http/client/GoodDataHttpClient.java

```diff
@@ -94,35 +92,72 @@
 
     private final SSTRetrievalStrategy sstStrategy;
 
-    private final HttpContext context;
+    /** Host performing authentication - eg. issuing TT tokens */
+    private final HttpHost authHost;
+
+    /** this lock is used to ensure that no threads will try to send requests while authentication is performed */
+    private final ReadWriteLock rwLock = new ReentrantReadWriteLock();
+
+    /** this lock guards that only one thread enters the authentication (obtaining TT/SST) section */
+    private final Lock authLock = new ReentrantLock();
+
+    /** current SST (or null if not yet obtained) */
+    private String sst;
+
+    /** TT to be set into the header (or null if not yet obtained) */
+    private String tt;
+
 
     /**
      * Construct object.
      * @param httpClient Http client
      * @param sstStrategy super-secure token (SST) obtaining strategy
      */
+    @Deprecated
```

Comment:
+1   
it's public library - it would be nice to put everything in javadoc.
Something like:
```java
/**
 * Construct object.
 * @param httpClient Http client
 * @param sstStrategy super-secure token (SST) obtaining strategy
 * @deprecated since 0.9.0 in favor of {@link #GoodDataHttpClient(HttpClient, HttpHost, SSTRetrievalStrategy)}
 */
```